modules/secrets/main.tf

```diff
@@ -0,0 +1,77 @@
+##############################################################################
+# Secret Group
+##############################################################################
+
+locals {
+  secret_groups = flatten([
+    for secret_group in var.secrets :
+    secret_group.existing_secret_group || secret_group.secret_group_name == null ? [] : [{
+      secret_group_name        = secret_group.secret_group_name
+      secret_group_description = secret_group.secret_group_description
+    }] if secret_group.secret_group_name != null
+  ])
+}
+
+data "ibm_sm_secret_groups" "secret_groups" {
+  instance_id   = var.existing_sm_instance_guid
+  region        = var.existing_sm_instance_region
+  endpoint_type = var.endpoint_type
+}
+
+module "secrets_manager_groups" {
+  for_each                 = { for obj in local.secret_groups : obj.secret_group_name => obj }
+  source                   = "terraform-ibm-modules/secrets-manager-secret-group/ibm"
+  version                  = "1.2.2"
+  region                   = var.existing_sm_instance_region
+  secrets_manager_guid     = var.existing_sm_instance_guid
+  secret_group_name        = each.value.secret_group_name
+  secret_group_description = each.value.secret_group_description
+  endpoint_type            = var.endpoint_type
+}
+
+##############################################################################
+# Secrets
+##############################################################################
+
+locals {
+  secrets = flatten([
+    for secret_group in var.secrets :
+    secret_group.existing_secret_group ? [
+      for secret in secret_group.secrets != null ? secret_group.secrets : [] : merge({
+        secret_group_id   = secret_group.secret_group_name != null ? data.ibm_sm_secret_groups.secret_groups.secret_groups[index(data.ibm_sm_secret_groups.secret_groups.secret_groups[*].name, secret_group.secret_group_name)].id : null
+        secret_group_name = secret_group.secret_group_name != null ? secret_group.secret_group_name : "default"
+      }, secret)
+      ] : [
+      for secret in secret_group.secrets != null ? secret_group.secrets : [] : merge({
+        secret_group_id   = secret_group.secret_group_name != null ? module.secrets_manager_groups[secret_group.secret_group_name].secret_group_id : null
+        secret_group_name = secret_group.secret_group_name != null ? secret_group.secret_group_name : "default"
```

Comment:
secret_group.secret_group_name is a required input, so it should never be null.

It is unclear what the intention is with the `default` secret group in this code. The behaviour is not documented in the readme. Do the lower level modules even support passing the id as default in this way to create secrets in the default secret group? I think maybe the secret_group_id is intended to default to be `default` (overloaded). It may be worth including a test for this code path.

Maybe a bigger question, secret_groups are the way to provide fine level access control over secrets. The use of default group may be considered a red flag for regulated environments, such as fscloud. Maybe default should just not be supported. @alex-reiff ?

Reply:
![Screenshot 2024-07-03 at 11 43 38 AM](https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/assets/122446118/36bd0b22-a6bd-4cbe-abe2-de978b735f2d)

Yes, it supports passing `default` as `secret_group_id`.

This submodule was added to support managing secrets created by the DA. [Issue](https://github.ibm.com/goldeneye/issues/issues/9801), so I think it should add flexibility for choosing which secret group to use.